internal/cutil/buffergroup.go

```diff
@@ -0,0 +1,84 @@
+package cutil
+
+// #include <stdlib.h>
+import "C"
+
+import (
+	"unsafe"
+)
+
+// BufferGroup is a helper structure that holds Go-allocated slices of
+// C-allocated strings and their respective lengths. Useful for C functions
+// that consume byte buffers with explicit length instead of null-terminated
+// strings. When used as input arguments in C functions, caller must make sure
+// the C code will not hold any pointers to either of the struct's attributes
+// after that C function returns.
+type BufferGroup struct {
+	// C-allocated strings.
+	CStrings []CharPtr
+	// Lengths of C strings, where Lengths[i] = length(CStrings[i]).
+	Lengths []SizeT
+}
+
+// NewBufferGroupStrings returns new BufferGroup constructed from Go strings.
+func NewBufferGroupStrings(strs []string) *BufferGroup {
+	s := &BufferGroup{
+		CStrings: make([]CharPtr, len(strs)),
+		Lengths:  make([]SizeT, len(strs)),
+	}
+
+	for i, str := range strs {
+		s.CStrings[i] = CharPtr(C.CString(str))
```

Comment:
I think `NewBufferGroupStrings()` needs to use `C.CBytes()` instead of `C.CString()`. The whole point of `BufferGroup` is to allow passing strings (buffers) of arbitrary contents (`\0` included) to C functions.

Functions like `rados.ReadOp.GetOmapValuesByKeys()`, `rados.WriteOp.SetOmap()` etc. still receive keys as `[]string` (and this will not change because it would be backwards incompatible). However, what we are essentially doing in this PR is allowing users to pass in arbitrary keys -- as if they were `[]byte`, not `string`. Using `C.CString()` breaks this premise, because it terminates the string on a null character.

And indeed, when I added `"ab\x00c"` into `TestBufferGroup/ValidContents` unit test, it failed:

```
=== RUN   TestBufferGroup/ValidContents
    buffergroup_test.go:49: 
        	Error Trace:	buffergroup_test.go:49
        	Error:      	Not equal: 
        	            	expected: "ab\x00c"
        	            	actual  : "ab"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-abc
        	            	+ab
        	Test:       	TestBufferGroup/ValidContents
```

Reply:
Yeah, good point!👍